test/parallel_api/ranges/std_ranges_test.h

```diff
@@ -0,0 +1,455 @@
+    {
+        assert(vec.size() == n);
+    }
+    type& operator()()
+    {
+        return vec;
+    }
+    ~usm_vector()
+    {
+        std::copy_n(vec.begin(), vec.size(), p);
+    }
+};
+
+template<typename T, typename ViewType>
+struct usm_subrange_impl
+{
+    using shared_allocator = sycl::usm_allocator<T, sycl::usm::alloc::shared>;
+    using type = ViewType;
+
+    shared_allocator alloc;
+    T* p = NULL;
+    ViewType view;
+
+    template<typename Policy>
+    usm_subrange_impl(Policy&& exec, T* data, int n): alloc(exec.queue()), p(data)
+    {
+        auto mem = alloc.allocate(n);
+        view = ViewType(mem, mem + n);
+        std::copy_n(data, n, view.data());
+    }
+
+    ViewType& operator()()
+    {
+        return view;
+    }
+
+    ~usm_subrange_impl()
+    {
+        std::copy_n(view.data(), view.size(), p);
+        alloc.deallocate(view.data(), view.size());
+    }
+};
+
+template<typename T>
+using  usm_subrange = usm_subrange_impl<T, std::ranges::subrange<T*>>;
+
+template<typename T>
+using  usm_span = usm_subrange_impl<T, std::span<T>>;
+
+#endif // _ONEDPL_HETERO_BACKEND
+
+template<typename T = int, TestDataMode TestDataMode = data_in, bool RetTypeCheck = true, bool ForwardRangeCheck = false>
+struct test_range_algo
+{
+    void operator()(auto algo, auto checker, auto... args)
+    {
+
+        auto subrange_view = [](auto&& v) { return std::ranges::subrange(v); };
+        auto span_view = [](auto&& v) { return std::span(v); };
+
+        if constexpr(ForwardRangeCheck)
+        {
+            auto forward_view = [](auto&& v) {
+                using forward_it = TestUtils::ForwardIterator<decltype(v.begin()), ::std::forward_iterator_tag>;
+                return std::ranges::subrange(forward_it(v.begin()), forward_it(v.end()));
+            };
+            test<T, host_vector<T>, TestDataMode, RetTypeCheck>{}(host_policies(), algo, checker, forward_view, args...);
+        }
+
+        test<T, host_vector<T>, TestDataMode, RetTypeCheck>{}(host_policies(), algo, checker, subrange_view, args...);
+        test<T, host_vector<T>, TestDataMode, RetTypeCheck>{}(host_policies(), algo, checker,  span_view, args...);
+        test<T, host_vector<T>, TestDataMode, RetTypeCheck>{}(host_policies(), algo, checker, std::views::all, args...);
+        test<T, host_subrange<T>, TestDataMode, RetTypeCheck>{}(host_policies(), algo, checker, std::views::all, args...);
+        test<T, host_span<T>, TestDataMode, RetTypeCheck>{}(host_policies(), algo, checker, std::views::all, args...);
+
+#if _ONEDPL_HETERO_BACKEND
+        //Skip the cases with pointer-to-function and hetero policy because pointer-to-function is not supported within kernel code.
+        if constexpr(!std::disjunction_v<std::is_member_function_pointer<decltype(args)>...>)
+        {
+            test<T, usm_vector<T>, TestDataMode, RetTypeCheck>{}(dpcpp_policy(), algo, checker, std::identity{},  args...);
```

Comment:
Can I host-side range be passed to the algorithm with the device policy? If yes, I suppose at least one case of it should be tested.

Reply:
No, you cannot. Support data with device policies is not a part MVP plan.

Reply:
I see, thanks